File: javascript/src/fp.js
/**
 * Apply a function on all elements. It can also be used on objects for which it
 * will map on object values but keep keys.
 * @param {Function} cb Function to apply on each elements
 * @returns {Function}
 */
const map = cb => d => Array.isArray(d) ?
  d.map(cb)
  : Object.keys(d).reduce((p, c) => Object.assign(p, { [c]: cb(d[c]) }), {});

/**
 * Reduce an array or an object.
 * @param {Function} cb Function to use when reducing the object
 * @param {any} start Accumulator default state
 * @returns {Function}
 */
const reduce = (cb, start) => d => Array.isArray(d) ?
  d.reduce(cb, start || [])
  : Object.keys(d).reduce((p, c) => cb(p, d[c], c), start || {});

/**
 * Flatten an array or an object values.
 * @param {any} d Array or object to flatten
 * @returns {Array}
 */
const flatten = d => Array.isArray(d) ?
  d.reduce((p, c) => p.concat(c), [])
  : flatten(Object.values(d));

/**
 * Mostly used with reduce, append one or more elements to an array or an object.
 * @param {...any} v Values to append
 * @returns {Object|Array}
 */
const append = (...v) => d => Array.isArray(d) ?
  d.concat(...v)
  : Object.assign(d, ...v);

/**
 * Clone an object.
 * @param {Object} o Object to clone
 * @returns {Object}
 */
const clone = o => Object.assign({}, o);

/**
 * Always returns the given value, used mostly for initialisation.
 * @param {any} v Value to return
 * @returns {Function}
 */
const always = v => () => v;

/**
 * Retrieve an object property.
 * @param {String} p Property to retrieve
 * @returns {Function} 
 */
const prop = p => d => d[p];

/**
 * Filter an array.
 * @param {Function} cb Function used for filtering
 * @returns {Function}
 */
const filter = cb => d => d.filter(cb);

/**
 * Instantiate a class.
 * @param {Function} klass Class ctor to use
 * @param  {...any} args Arguments to use
 * @returns {Function}
 */
const instantiate = (klass, ...args) => d => new klass(d, ...args);

/**
 * Retrieve the first element of an array.
 * @param {Array} d The array
 * @returns {Function}
 */
const first = d => d[0];

/**
 * Retrieve the last element of an array.
 * @param {Array} d The array
 * @returns {Function}
 */
const last = d => d[d.length - 1];

/**
 * Returns a function which pipe the given ones.
 * @param  {...any} functions Functions to chain
 * @returns {Function}
 */
const pipe = (...functions) => args => functions.reduce((arg, fn) => fn(arg), args);

module.exports = {
  append,
  clone,
  first,
  last,
  map,
  filter,
  reduce,
  flatten,
  always,
  prop,
  instantiate,
  pipe,
};

File: python/pychatl/fp.py
import functools, itertools

def map(func):
  """Apply a function on all elements. It can also be used on objects for which 
  it will map on object values but keep keys.

  Args:
    func (callable): Callback to apply on each elements

  """
  def cb(d):
    if isinstance(d, list):
      return [ func(v) for v in d]
    else:
      return { k: func(v) for k, v in sorted(d.items(), key=lambda x: x[0]) }
  return cb

def reduce(func, start=None):
  """Reduce an array or an object.

  Args:
    func (callable): Function to use when reducing the object.
    start (any): Accumulator default state

  """
  def cb(d):
    if isinstance(d, list):
      return functools.reduce(func, d,
        start if start is not None else [])
    else:
      return functools.reduce(lambda p, c: func(p, d[c], c), d.keys(),
        start if start is not None else {}) 
  return cb

def clone(o):
  """Clone an object.

  Args:
    o (dict): Dictionary to clone
  
  """
  return o.copy()

def always(value):
  """Always returns the given value, used mostly for initialisation.

  Args:
    value (any): Value to return

  """
  def cb(d):
    return value
  return cb

def prop(name):
  """Retrieve an object property.

  Args:
    name (str): Name of the property to retrieve
  
  """
  def cb(o):
    return o.get(name)
  return cb

def filter(func):
  """Filter an array.

  Args:
    func (callable): Function for filtering

  """
  def cb(d):
    return [v for v in d if func(v)]
  return cb

def flatten(d):
  """Flatten an array or an object values.

  Args:
    d (list or dict): List or dict to flatten

  """
  seq = d if isinstance(d, list) else d.values()
  return functools.reduce(lambda p, c: p + c, seq, [])

def append(*v):
  """Mostly used with reduce, append one or more elements to an array or an object by mutating it.

  Args:
    v (list): Values to append

  """
  def cb(d):
    if isinstance(d, list):
      return d.extend(list(itertools.chain(v))) or d
    else:
      return d.update(dict(itertools.chain(*(i.items() for i in v)))) or d
  return cb

def instantiate(klass, *args):
  """Instantiate a class.

  Args:
    klass (class): Class to instantiate
    args (any): Args to provide to the constructor along with the current value

  """
  def cb(d):
    return klass(d, *args)
  return cb

def pipe(*functions):
  """Returns a function which pipe the given ones.

  Args:
    functions (list of callable): Functions to chain

  """
  def cb(d):
    return functools.reduce(lambda arg, func: func(arg), functions, d)

  return cb